pgrx/src/memcx.rs

```diff
@@ -16,27 +18,35 @@ pub struct MemCx<'mcx> {
 }
 
 impl<'mcx> MemCx<'mcx> {
-    /// You probably shouldn't be using this.
+    /// Wrap the provided [`pg_sys::MemoryContext`]
+    ///
+    /// # Safety:
+    /// Assumes the provided [`pg_sys::MemoryContext`] is valid and properly initialized.
+    /// This method does check to ensure the pointer is non-null, but that is the only sanity
+    /// check that is performed.
     pub(crate) unsafe fn from_ptr(raw: pg_sys::MemoryContext) -> MemCx<'mcx> {
         let ptr = NonNull::new(raw).expect("memory context must be non-null");
         MemCx { ptr, _marker: PhantomData }
     }
 
-    /// You probably shouldn't be using this.
-    pub(crate) unsafe fn alloc_bytes(&self, size: usize) -> *mut u8 {
+    /// Allocate a raw byte buffer `size` bytes in length
+    /// and returns a pointer to the new allocation.
+    pub unsafe fn alloc_bytes(&self, size: usize) -> *mut u8 {
         pg_sys::MemoryContextAlloc(self.ptr.as_ptr(), size).cast()
     }
 
-    /// You probably shouldn't be using this.
-    pub(crate) unsafe fn exec_in<T>(&self, f: impl FnOnce() -> T) -> T {
+    /// Stores the current global memory context, switches to *this* memory context,
+    /// and executes the closure `f`.
+    /// Once `f` completes, the previous global memory context is restored.
+    pub unsafe fn exec_in<T>(&self, f: impl FnOnce() -> T) -> T {
         let remembered = pg_sys::MemoryContextSwitchTo(self.ptr.as_ptr());
         let res = f();
```

Comment:
if `f()` panics then we're left in `self`'s memory context and not whatever we `remembered`.  I'm not sure this is what we'd want here.  It might make sense to use `PgTryBuilder()` to run `f()`, which would give us a "finally" block where we can make sure to restore the memory context to `remembered`.

Reply:
This is not meant to be commonly used.

An additional "this is UB if anything panics" constraint is acceptable.

Reply:
> An additional "this is UB if anything panics" constraint is acceptable.

Fine by me.  ;)

Reply:
Per Discord conversation on correctness, documented this as a safety constraint instead in https://github.com/pgcentralfoundation/pgrx/pull/1737/commits/f149072c318f7e39c212daf3b59e0f723d1dfd46